SLS-1448 Fix layered cursor next/prev scans if checkpoint changes in the middle

ddanderson opened the pull request:
- Extended test_layered31 to cover this situation (for next, at least).
- The position has to be copied from the old stable cursor to the new stable cursor, it's easier to do this via a reopen.
- Some refactoring to put the stable open into its own function, since that stable opens are now done in multiple places.
- A small bug fix (flag snafu) and some test code cleanup.

ddanderson commented:
@pmacko86 @korteland This is ready for review, it's a bit bigger change, see the PR comment above.  I was also thinking ahead to SLS-1449 (step-down step-up) and I thought clarifying the state change code was helpful.

@agorrod, I didn't mark you as a reviewer, but you're welcome to take a look.

agorrod reviewed (approved):
I skimmed the change and it makes a lot of sense to me.

korteland reviewed (commented):
I really like the overall design of this - nice work!

pmacko86 commented:
@ddanderson, I have a meta-level question: How does this interplay with the snapshot isolation semantics? I see that a single cursor scan would be able to see values from two different database states—is this something that we allow in classic WT?

At which point does WT grab the transaction snapshot if the transaction has not been explicitly started—is it at the first `next`, or at each `next`? Just wanting to make sure that we don't change semantics with disagg, as that would have implications on mongod correctness.

Even if we do, we should have a test that does the same thing, except that the follower does a begin transaction before the scan and rollback afterwards, NOT expecting to see the new values.

ddanderson commented:
@pmacko86 That's a great meta-question, I just looked it up in the WT doc [here](https://source.wiredtiger.com/develop/transactions_api.html). I see:

At read-committed isolation, committed changes from concurrent transactions become visible only when no cursor is positioned. In other words, at this isolation level, all cursors in a session read from a stable snapshot while any cursor in the session remains positioned. A call to [WT_CURSOR::next](https://source.wiredtiger.com/develop/struct_w_t___c_u_r_s_o_r.html#a0503f16bd8f3d05aa3552f229b3a8e1b) or [WT_CURSOR::prev](https://source.wiredtiger.com/develop/struct_w_t___c_u_r_s_o_r.html#a43d6664d2f68902aa63f933864242e76) on a positioned cursor will not update the snapshot.

At snapshot isolation when using implicit transactions, the same behavior applies for reads: when a read is performed, a snapshot is taken, and all cursors will read from that snapshot as long as any cursor remains positioned or until a write operation is performed. Further reads after that point (or a call to [WT_SESSION::reset_snapshot](https://source.wiredtiger.com/develop/struct_w_t___s_e_s_s_i_o_n.html#a2117021276bad0e6f541e61a5459ec59)) will acquire a new snapshot.

=-=-=-=-=-=

So there are some guarantees, and in the _general_ case perhaps the safest thing is to not upgrade the cursor, except perhaps reopening a cached cursor when it gets reused. That will never give the wrong answer, but it doesn't free up cursor references like we want.  I think Mongodb always (or almost always?) uses a read_timestamp transaction.  If you have a read_timestamp transaction it's always legal to upgrade (even during next/prev), you'll get the same consistent view at the next checkpoint (courtesy of history store).

It looks pretty cheap to check if there's a read_timestamp, so I'll change to only upgrade if that is set.  It's quite possible that layered data stores are only to be used with read timestamped transactions.  I'll check that out with the mongod folks, if it's the case, we can return some bad error if there's not a timestamp.

pmacko86 commented:
Thank you for checking, this makes sense, and yes, please check about the read timestamp. My preference would be to keep reading without timestamps allowed but unoptimized, at least for now, as most of our tests don't use timestamps - and as prohibiting it could create another dependency when bringing this to develop a feature at a time as Will's document suggested.

ddanderson commented on src/cursor/cur_layered.c:
```diff
@@ -229,19 +312,66 @@ __clayered_adjust_state(WT_CURSOR_LAYERED *clayered, bool *state_updated)
+             * we copy the position.
+             */
+            old_stable = clayered->stable_cursor;
+            clayered->stable_cursor = NULL;
```
clayered is a cursor (WT_CURSOR_LAYERED), and cursors are specific to a session, there's only one thread that can legally access them at a time.  I'm not sure this is what you are thinking of, but WT_LAYERED is different (it's a kind of dhandle), that is multi-threaded - but the stuff in the WT_LAYERED specific part is created when a layered table's dhandle is first opened, and never changed.

korteland replied:
Thanks - for some reason I had it in my head that the connection-level layered table manager had some admin jobs related to these cursors. But you're right, only one thread at a time can legally access these.

ddanderson commented on src/cursor/cur_layered.c:
```diff
@@ -206,6 +206,84 @@ __clayered_close_cursors(WT_CURSOR_LAYERED *clayered, bool include_ingest)
+ */
+static int
+__clayered_open_stable(
+  WT_CURSOR_LAYERED *clayered, bool leader, const char **cfg, u_int cfg_len, u_int cfg_pos)
```
__clayered_open_stable is going to need the entire cfg array, since it uses it to open.

korteland replied:
Thanks - I didn't realise.

korteland commented on src/cursor/cur_layered.c:
```diff
@@ -229,19 +312,66 @@ __clayered_adjust_state(WT_CURSOR_LAYERED *clayered, bool *state_updated)
+
+        /* A random stable cursor shouldn't be reopen, it may have additional state. */
+        if (clayered->stable_cursor == NULL || F_ISSET(clayered, WT_CLAYERED_RANDOM))
+            reopen_stable = false;
```
Oh I see - that makes sense. Thanks

quchenhao commented on src/cursor/cur_layered.c:
```diff
@@ -229,19 +312,66 @@ __clayered_adjust_state(WT_CURSOR_LAYERED *clayered, bool *state_updated)
     else
         current_checkpoint_id = 0;
```
Change 0 to WT_DISAGG_CHECKPOINT_ID_NONE?

ddanderson replied:
Thanks, fixed!

quchenhao commented on src/cursor/cur_layered.c:
```diff
@@ -229,19 +312,66 @@ __clayered_adjust_state(WT_CURSOR_LAYERED *clayered, bool *state_updated)
+
+        if (current_leader != clayered->leader) {
+            /* If leader state has changed, all cursors are affected. */
+            reopen_ingest = true;
```
I wonder when do we need to reopen the ingest table?

ddanderson replied:
For the ingest table, we're either going to close it (step up) or open it (step down).  `reopen_ingest` actually handles both of those.  I realized that "reopen_ingest" is not a great name, so changed it to `change_ingest`.

As for the stable cursor, there are some subtleties about what should happen during step up and step down, I'm deferring resolution of this until SLS-1449.

quchenhao commented on src/cursor/cur_layered.c:
```diff
@@ -229,19 +312,66 @@ __clayered_adjust_state(WT_CURSOR_LAYERED *clayered, bool *state_updated)
+
+        /* See if there's nothing to do for the ingest cursor. */
+        if (clayered->ingest_cursor == NULL)
+            reopen_ingest = false;
```
I think we should never see a layered table with NULL ingest cursor?

quchenhao commented:
@ddanderson With this change, is this still necessary to pin content in the ingest table? How do we ensure we can read the data from the updated stable table?

ddanderson commented:
@quchenhao,
> With this change, is this still necessary to pin content in the ingest table? How do we ensure we can read the data from the updated stable table with the current read timestamp?

I think we still need to.  This PR is still in progress, and in particular, we won't be able to upgrade every layered cursor immediately.  So there will still be cases where we need the data in the ingest table, because we are still reading at an older stable table.

ddanderson commented on src/cursor/cur_layered.c:
```diff
@@ -206,6 +206,84 @@ __clayered_close_cursors(WT_CURSOR_LAYERED *clayered, bool include_ingest)
+    if (!leader) {
+        /*
+         * We may have a stable chunk with no checkpoint yet. If that's the case then open a cursor
+         * on stable without a checkpoint. It will never return an invalid result (it's content is
```
Fixed

ddanderson commented on test/suite/test_layered31.py:
```diff
@@ -195,7 +195,79 @@ def test_layered31(self):
+            cursor = follower_cursors[uri]
+            found = first_read
+            for i in range(first_read, self.nitems):
+                if cursor.next() != 0:
```
Thanks, fixed!

ddanderson commented on src/cursor/cur_layered.c:
```diff
@@ -206,6 +206,84 @@ __clayered_close_cursors(WT_CURSOR_LAYERED *clayered, bool include_ingest)
+            cfg[cfg_pos++] = "readonly";
+        }
+    }
+    WT_ASSERT_ALWAYS(session, cfg_pos < cfg_len, "configuration array too small");
```
I rewrote this a bit differently, it gets rid of cfg_pos.

korteland replied:
Thanks - I find the new version easier to read as well, fwiw.

ddanderson commented:
@quchenhao, @pmacko86, @korteland, @agorrod, I've addressed the review comments and:
- streamlined the config arg array stuff.
- never reopen during cursor iteration operation (next/prev)
- don't reopen stable unless:
-- using a read timestamp
-- session currently not using a snapshot
-- snapshot has changed

The goal is to be compatible with [our documentation on transactions](https://source.wiredtiger.com/develop/transactions_api.html), and at the same time, close old stable cursors promptly so we're not pinning data in the ingest table.

Some of this may be wrong - you all probably know a lot more about transactional semantics than I do.

quchenhao commented on src/cursor/cur_layered.c:
```diff
@@ -206,42 +206,229 @@ __clayered_close_cursors(WT_CURSOR_LAYERED *clayered, bool include_ingest)
-
+    session = CUR2S(clayered);
+    conn = S2C(session);
+    current_leader = conn->layered_table_manager.leader;
```
nitpick: I think we can only change leadership when no transaction is running so it is safe to read it this way. But it is better to use a atomic relaxed read to read it as leader is a shared variable.

ddanderson replied:
I'm not familiar with "atomic relaxed read". Are we talking about this?
```
        WT_ACQUIRE_READ(current_leader, conn->layered_table_manager.leader);
```
I'm in favor of doing things the safer way, though I don't see other places in the code that get the global "leader" variable, except as a naked read.

quchenhao replied:
I mean we can use __wt_atomic_load8. I understand why are not using that for now. So we can ignore it for now. However, we do need to clean it up some time in the future otherwise TSAN build will fail.

ddanderson replied:
Got it, thanks!

quchenhao commented on src/cursor/cur_layered.c:
```diff
@@ -1006,7 +1129,7 @@ __clayered_search(WT_CURSOR *cursor)
     WT_ERR(__cursor_needkey(cursor));
     __cursor_novalue(cursor);
-    WT_ERR(__clayered_enter(clayered, true, false));
+    WT_ERR(__clayered_enter(clayered, true, false, false));
```
Why not updating the checkpoints when searching?

ddanderson replied:
This is in the transaction doc:  "A call to [WT_CURSOR::next](https://source.wiredtiger.com/develop/struct_w_t___c_u_r_s_o_r.html#a0503f16bd8f3d05aa3552f229b3a8e1b) or [WT_CURSOR::prev](https://source.wiredtiger.com/develop/struct_w_t___c_u_r_s_o_r.html#a43d6664d2f68902aa63f933864242e76) on a positioned cursor will not update the snapshot."  The strategy used here is: when the snapshot can be (or is) updated, then we can update the stable cursor.

quchenhao replied:
But are we doing the opposite, we only update the stable cursor when the snapshot cannot be updated?

quchenhao replied:
false means updating the checkpoint actually.

quchenhao reviewed (approved):
lgtm

quchenhao commented on src/cursor/cur_layered.c:
```diff
@@ -206,42 +206,229 @@ __clayered_close_cursors(WT_CURSOR_LAYERED *clayered, bool include_ingest)
+            /*
+             * Layered cursors are sometimes paired with read timestamps. When using read
+             * timestamps, it's always safe to update cursors, even during iterations. That's
+             * because the view at a timestamp is always consistent, the history store covers that.
```
So why we disable update for iteration even with read timestamp?

ddanderson replied:
Ah thanks!  It should always allow updates with read timestamps.  I'll fix.

korteland reviewed (approved):
LGTM!
